beluga/include/beluga/actions/normalize.hpp

```diff
@@ -34,120 +35,106 @@ namespace beluga::actions {
 
 namespace detail {
 
-/// Implementation detail for a normalize range adaptor object.
-struct normalize_base_fn {
-  /// Overload that implements the normalize algorithm.
-  /**
-   * \tparam ExecutionPolicy An [execution policy](https://en.cppreference.com/w/cpp/algorithm/execution_policy_tag_t).
-   * \tparam Range An [input range](https://en.cppreference.com/w/cpp/ranges/input_range).
-   * \param policy The execution policy to use.
-   * \param range An existing range to apply this action to.
-   * \param factor The normalization factor.
-   */
-  template <
-      class ExecutionPolicy,
-      class Range,
-      std::enable_if_t<std::is_execution_policy_v<std::decay_t<ExecutionPolicy>>, int> = 0,
-      std::enable_if_t<ranges::range<Range>, int> = 0>
-  constexpr auto operator()(ExecutionPolicy&& policy, Range& range, double factor) const -> Range& {
-    if (std::abs(factor - 1.0) < std::numeric_limits<double>::epsilon()) {
-      return range;  // No change.
-    }
+/// \cond detail
+
+template <class ExecutionPolicy = std::execution::sequenced_policy>
+struct normalize_closure {
+ public:
+  static_assert(std::is_execution_policy_v<ExecutionPolicy>);
+
+  constexpr normalize_closure() noexcept : policy_{std::execution::seq} {}
+
+  constexpr explicit normalize_closure(ExecutionPolicy policy) : policy_{std::move(policy)} {}
+
+  constexpr explicit normalize_closure(double factor) noexcept : policy_{std::execution::seq}, factor_{factor} {}
 
-    auto weights = [&range]() {
+  constexpr normalize_closure(ExecutionPolicy policy, double factor) : policy_{std::move(policy)}, factor_{factor} {}
+
+  template <class Range>
+  constexpr auto operator()(Range& range) const -> Range& {
+    static_assert(ranges::forward_range<Range>);
+
+    auto weights = std::invoke([&range]() {
       if constexpr (beluga::is_particle_range_v<Range>) {
         return range | beluga::views::weights | ranges::views::common;
       } else {
         return range | ranges::views::common;
       }
-    }();
+    });
+
+    const double factor = std::invoke([this, weights]() {
+      if (factor_.has_value()) {
+        return factor_.value();
+      }
+
+      return ranges::accumulate(weights, 0.0);  // The default normalization factor is the total sum of weights.
+    });
+
+    if (std::abs(factor - 1.0) < std::numeric_limits<double>::epsilon()) {
+      return range;  // No change.
+    }
 
     std::transform(
-        policy,               //
-        std::begin(weights),  //
-        std::end(weights),    //
-        std::begin(weights),  //
+        policy_,          //
+        weights.begin(),  //
+        weights.end(),    //
+        weights.begin(),  //
         [factor](const auto w) { return w / factor; });
```

Comment:
meta: Should we add some kind of `assert` to ensure that factor won't take a zero value? Probably when weights is empty.